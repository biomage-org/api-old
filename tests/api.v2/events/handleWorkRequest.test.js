const handleWorkRequest = require('../../../src/api.v2/events/handleWorkRequest');
const validateAndSubmitWork = require('../../../src/api.v2/events/validateAndSubmitWork');
const fake = require('../../test-utils/constants');

jest.mock('../../../src/api.v2/middlewares/authMiddlewares');
jest.mock('aws-xray-sdk');

jest.mock('../../../src/api.v2/events/validateAndSubmitWork', () => jest.fn());

const data = {
  uuid: 'someuuid-asd-asd-asdddasa',
  Authorization: 'Bearer some-token',
  experimentId: fake.EXPERIMENT_ID,
};

describe('Handle work socket callback', () => {
  const mockSocket = { emit: jest.fn() };
  const mockXraySegment = { addError: jest.fn() };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('submits work', async () => {
    validateAndSubmitWork.mockImplementation(() => Promise.resolve());

    await handleWorkRequest(mockSocket, data, mockXraySegment);
    expect(mockSocket.emit.mock.calls[0]).toMatchSnapshot();
    expect(validateAndSubmitWork.mock.calls[0]).toMatchSnapshot();
  });

  it('If the worker fails, the API emits a work-response with error', async () => {
    validateAndSubmitWork.mockImplementation(() => Promise.reject({ message: 'fail' }));

    await handleWorkRequest(mockSocket, data, mockXraySegment);
    expect(mockSocket.emit.mock.calls[0]).toMatchSnapshot();
    expect(mockXraySegment.addError).toHaveBeenCalled();
    expect(validateAndSubmitWork.mock.calls[0]).toMatchSnapshot();
  });

  it('data without authorization also throws an error', async () => {
    delete data.Authorization;
    await handleWorkRequest(mockSocket, data, mockXraySegment);
    expect(mockSocket.emit.mock.calls[0]).toMatchSnapshot();
  });
});
