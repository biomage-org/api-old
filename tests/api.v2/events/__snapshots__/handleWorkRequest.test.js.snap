// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle work socket callback If the worker fails, the API emits a work-response with error 1`] = `
Array [
  "WorkResponse-someuuid-asd-asd-asdddasa",
  Object {
    "request": Object {
      "Authorization": "Bearer some-token",
      "experimentId": "experi-mentid-11111111-1111111-11111",
      "uuid": "someuuid-asd-asd-asdddasa",
    },
    "response": Object {
      "cacheable": false,
      "error": "fail",
      "trace": undefined,
    },
    "results": Array [],
  },
]
`;

exports[`Handle work socket callback If the worker fails, the API emits a work-response with error 2`] = `
Array [
  Object {
    "Authorization": "Bearer some-token",
    "experimentId": "experi-mentid-11111111-1111111-11111",
    "uuid": "someuuid-asd-asd-asdddasa",
  },
]
`;

exports[`Handle work socket callback data without authorization also throws an error 1`] = `
Array [
  "WorkResponse-someuuid-asd-asd-asdddasa",
  Object {
    "request": Object {
      "experimentId": "experi-mentid-11111111-1111111-11111",
      "uuid": "someuuid-asd-asd-asdddasa",
    },
    "response": Object {
      "cacheable": false,
      "error": "Authentication token must be present.",
      "trace": undefined,
    },
    "results": Array [],
  },
]
`;

exports[`Handle work socket callback submits work 1`] = `
Array [
  "WorkerInfo-experi-mentid-11111111-1111111-11111",
  Object {
    "response": Object {
      "podInfo": undefined,
      "trace": undefined,
    },
  },
]
`;

exports[`Handle work socket callback submits work 2`] = `
Array [
  Object {
    "Authorization": "Bearer some-token",
    "experimentId": "experi-mentid-11111111-1111111-11111",
    "uuid": "someuuid-asd-asd-asdddasa",
  },
]
`;
